### ISSUE TITLE
Some packages scanned but license was not discovered although the license file exists

### ISSUE THREAD

**Issue by RabeeaEgbareia** [nexB/scancode.io#662]
I scanned a zip file with several packages with scan_codebase pipeline, the job finished successfully but I see some packages(for example: iperf package) was scanned but no license discovered although the licenses file (LICENSE) exists.
I'm using the api: /api/projects/[PROJECT_ID]/packages/ to get the packages license

I attached the example zip file I mentioned above (iperf.zip) 

* I'm using v32.10 
![image](https://user-images.githubusercontent.com/128488364/229861293-f0006d34-7291-4c76-9485-9fe0078239ed.png)

[iperf.zip](https://github.com/nexB/scancode.io/files/11150914/iperf.zip)

**Comment by mjherzog** [comment]
iperf.zip is an archive, not a Package in the ScanCode.io context. SCIO will only report package data where you have package metadata as for npms, RubyGems, etc. You should be able to get the file-level data for iperf from the Scan Resources.

**Comment by RabeeaEgbareia** [comment]
@mjherzog 
So seems I misunderstand something ..
1. "SCIO will only report package data where you have package metadata as for npms, RubyGems, etc.", where can I find all the list ?
2. what do you suggest me to do to find the license(s) used in iperf.zip ? I see scan resources results and it a lot of information. Can you please explain what should I take ? 
I'm using the api: api/projects/[PROJECT_ID]/results/

**Comment by tdruez** [comment]
> "SCIO will only report package data where you have package metadata as for npms, RubyGems, etc.", where can I find all the list ?

https://scancode-toolkit.readthedocs.io/en/doc-update-licenses/reference/available_package_parsers.html

> what do you suggest me to do to find the license(s) used in iperf.zip ?

1. You can have a look at the charts from the project page UI to get an overview of the licenses detected in the input archive:
![Screenshot 2023-04-05 at 11 59 35](https://user-images.githubusercontent.com/489057/230018531-55b1a104-da79-44c4-989a-7ea31f7dd55f.jpg)


2. You can click on the chart to get a list of resources related to a specific license or check the whole list of resources.
![Screenshot 2023-04-05 at 12 02 43](https://user-images.githubusercontent.com/489057/230019443-7d556690-f39d-4e21-9a80-e81f7d31a385.jpg)


3. You can look at a file content to visualize snippets of text where the licenses where detected:
![Screenshot 2023-04-05 at 12 08 49](https://user-images.githubusercontent.com/489057/230020895-0d53be79-acfb-42cb-97ea-699fcc353c50.jpg)


> I see scan resources results and it a lot of information.

Indeed.

> I'm using the API

REST APIs are a great way to exchange data, not so much to review scan results.

**Comment by RabeeaEgbareia** [comment]
@mjherzog 
I'm integrating scancode.io as automation tool to my product. For the automation I'm using only the rest api of scancode.io (including uploading files and starting pipeline) and I need to report all the results with teamCity ( the final results for me are to know all the **licenses used** in my different projects and which **packages they belong to**).
So the graphs in this case are nice extra information and good to know but the graphs are not the goal.

> REST APIs are a great way to exchange data, not so much to review scan results. 
So what's you suggestion ? what's the fastest way to get my goal I mentioned above ?

**Comment by RabeeaEgbareia** [comment]
@tdruez 
Thanks a lot !

**Comment by RabeeaEgbareia** [comment]
@tdruez 
A question:
Why ["packages"] field has only one license expression ? and ["files"] could have several license expressions ? 
Is a package like rpm, npms, RubyGems, etc can have only one license and and archive folder (like iperf) can have more than one license  ?

**Comment by AyanSinhaMahapatra** [comment]
Hey @RabeeaEgbareia 

> TL;DR: this was a wart, for no specific reason. Being standardized in later versions with one license-expression string for both file/package.

This is scancode-toolkit v31.x output format btw, we are releasing v32.x shortly and we have a seperate output format where both in files and packages we have a single (detected/declared) license expression. (and a secondary license expression for packages)  We will also update scancode.io to this new output format, see #569 

Previously in v31 (which is what latest scancode.io is using currently) we `licenses` containing license matches and `license-expressions` with their respective `license-expression` and this was a list. But for packages we didn't show license matches similarly, and in most cases performed case-specific license detections and hence we had just one license-expression as a result. 

See also https://scancode-toolkit.readthedocs.io/en/doc-update-licenses/reference/license-detection-reference.html for more details on this.

**Comment by RabeeaEgbareia** [comment]
@AyanSinhaMahapatra 
Thanks you for the clarification. 
For now I will manage with the current output, I will update my project with the new release v32.x once it released.
Do you have an estimated time ? how can I be noticed/updated with the new release ?

**Comment by AyanSinhaMahapatra** [comment]
> Do you have an estimated time ?

I hope this week :crossed_fingers: 

> I will update my project with the new release v32.x once it released

Great! It has a lot of improvements.
But note that v32.0.0rc3 is also stable in terms of output format, we are not changing anything there anymore.

> how can I be noticed/updated with the new release ?

We do announce new releases in https://matrix.to/#/#aboutcode-org_discuss:gitter.im and https://matrix.to/#/#aboutcode-org_scancode:gitter.im, and we are also considering to automate release updates for all our projects: https://github.com/nexB/aboutcode/issues/122 but we are not there yet

**Comment by pombredanne** [comment]
> @tdruez A question: Why ["packages"] field has only one license expression ? and ["files"] could have several license expressions ? Is a package like rpm, npms, RubyGems, etc can have only one license and and archive folder (like iperf) can have more than one license ?

@RabeeaEgbareia The license_expression of a package is derived from a single input which is the package manifest field that stores a license, such as a Maven POM `<licenses>` tag or an npm '`license`" package.json attribute: we derive a single license expression from this single tag. 

In constrast, each file may have many different discrete license statements in multiple positions: we report one license expression for each in these cases.

Note that as @AyanSinhaMahapatra mentioned  above we have updated formats coming up.

@RabeeaEgbareia on another note you wrote:

>  For now I will manage with the current output, I will update my project with the new release v32.x once it released.

I am curious about what you project is! Can you share some details about it?

**Comment by RabeeaEgbareia** [comment]
@pombredanne 
Thanks ! so, do you suggest me to wait with my integration in my project with scancode.io until the new version be released ? Does the new version have significant changes in the format ?

@pombredanne for your second question  
> I am curious about what you project is! Can you share some details about it?

As I wrote above:

> I'm integrating scancode.io as automation tool to my product. For the automation I'm using only the rest api of scancode.io (including uploading files and starting pipeline) and I need to report all the results with teamCity ( the final results for me are to know all the licenses used in my different projects and which packages they belong to).
So the graphs in this case are nice extra information and good to know but the graphs are not the goal.

And:
I'm integrating scancode.io api in Java, which means, writing java code to integrate scancode.io api to scan projects we use in our company and to be aware which licenses we use in our projects and which packages use these licenses. 
My scan results and analyzes should look as the following:
![image](https://user-images.githubusercontent.com/128488364/231132517-3723870b-7a6b-4ff1-b7cb-6ff45972faea.png)

Every week we should scan our projects and know if we added new packages in use and which licenses added in that week and which licenses are already in use ("old licenses")

I hope I explained myself clear enough, if not, let me know if you have more questions.

**Comment by pombredanne** [comment]
>  Thanks ! so, do you suggest me to wait with my integration in my project with scancode.io until the new version be released ? Does the new version have significant changes in the format ?

I would not wait. We will try to minimize the impact and will provide upgrade instructions when this happens.

**Comment by RabeeaEgbareia** [comment]
Ok thanks

**Comment by RabeeaEgbareia** [comment]
Hi all, 
@pombredanne @tdruez @AyanSinhaMahapatra @mjherzog 
I have another question: 

What is the difference between "license_expression" and "declared_license" in "packages" ?
Why I'm getting such like this results? 
(I'm with the same images version)

"license_expression": "unknown",
"declared_license": "[{'url': 'http://www.apache.org/licenses/LICENSE-2.0.txt', 'name': 'Apache 2', 'comments': None, 'distribution': 'repo'}]",

**Comment by pombredanne** [comment]
@RabeeaEgbareia You wrote

> I have another question:

It is usually better to re start a new issue for a new topic ;)

> What is the difference between "license_expression" and "declared_license" in "packages" ?

- `"declared_license"` is the original license tag, field or structure found in a package manifest, converted to a plain data structure. For instance this `"declared_license": "[{'url': 'http://www.apache.org/licenses/LICENSE-2.0.txt', 'name': 'Apache 2', 'comments': None, 'distribution': 'repo'}]",` is likely seen in a Maven POM XML such as this https://repo1.maven.org/maven2/org/parboiled/parboiled-core/0.10.1/parboiled-core-0.10.1.pom, but here converted the XML to plain data.   Note that the "declared_license" field is renamed to "extracted_license_statement" in the upcoming release of ScanCode Toolkit and will be renamed also in ScanCode.io

- `"license_expression"` is the normalized license expression detected from the declared_license/extracted_license_statement field. It is a bug that we do report this as "unknown" here. There is a lot of pending fixes worked on Maven POM licenses ongoing in https://github.com/nexB/scancode-toolkit/pull/3346 FWIW.

**Comment by RabeeaEgbareia** [comment]
@pombredanne 
> It is usually better to re start a new issue for a new topic ;) 
Ok, Got it :) 

Thanks for the explanation. 
I used "license_expression" in my project, what do you suggest me to do ? Is there any estimation when this will be fixed ?
For now, seems not trivial (for me) to start using "declared_license" instead of "license_expression" ...

**Comment by AyanSinhaMahapatra** [comment]
@RabeeaEgbareia this is fixed in scancode-toolkit and we are working on https://github.com/nexB/scancode.io/issues/569 which would be in [scancode.io v33](https://github.com/nexB/scancode.io/milestone/6) here. I'm not sure I have an accurate estimate on the release, but this is being actively worked on and is top priority for us at this point.

As mentioned above, we have also streamlined and renamed some of our license fieldnames as [seen here in the CHANGELOG](https://github.com/nexB/scancode-toolkit/blob/develop/CHANGELOG.rst?plain=1#L120), and we would have detailed documentation on this too for upgrading.

> I used "license_expression" in my project, what do you suggest me to do 

This would be the way IMHO for now, as opposed to using `declared_license` which are not normalized/mapped to actual license expressions, just the text as is, as found in the manifest.

**Comment by RabeeaEgbareia** [comment]
Hi again, @AyanSinhaMahapatra 
Is there any progress regarding this issue ? estimation for the stable version ? 

> @RabeeaEgbareia this is fixed in scancode-toolkit and we are working on #569 which would be in [scancode.io v33](https://github.com/nexB/scancode.io/milestone/6) here. I'm not sure I have an accurate estimate on the release, but this is being actively worked on and is top priority for us at this point.
> 
> As mentioned above, we have also streamlined and renamed some of our license fieldnames as [seen here in the CHANGELOG](https://github.com/nexB/scancode-toolkit/blob/develop/CHANGELOG.rst?plain=1#L120), and we would have detailed documentation on this too for upgrading.
> 
> > I used "license_expression" in my project, what do you suggest me to do
> 
> This would be the way IMHO for now, as opposed to using `declared_license` which are not normalized/mapped to actual license expressions, just the text as is, as found in the manifest.

**Comment by AyanSinhaMahapatra** [comment]
Hi @RabeeaEgbareia 
Yes!

We do have a stable release of scancode-toolkit out now, currently at [32.0.4](https://github.com/nexB/scancode-toolkit/releases/tag/v32.0.4) and this is also supported in the latest scancode.io release: https://github.com/nexB/scancode.io/releases/tag/v32.3.0 (upgrade done in https://github.com/nexB/scancode.io/pull/752 and https://github.com/nexB/scancode.io/pull/772, see these and the documentation at https://scancode-toolkit.readthedocs.io/en/stable/reference/license-detection-reference.html for updating reference).

**Comment by RabeeaEgbareia** [comment]
@AyanSinhaMahapatra 
Thank you ! I will install it again and try.

**Comment by tdruez** [comment]
Closing as completed.
